WIP: New Constant string codec for strings longer than 8 bytes.

TomHodson opened the pull request:
WIP: FIxes https://github.com/ecmwf/pyodc/pull/14

## So far

Added a failing test that selects a codec for constant column with >8 byte encoded strings and trys to round trip encode/decode it. I may have misunderstood how to use the codecs directly in the test code though.

## Next steps

Add a new codec that allows encoding variable length constants using two fields in the header:
    - a uint64 `length` field 
    - a variable length byte field of size `length` that encodes the constant data for that column

Currently there is a codec `Constant` which encode a constant value up to 8 bytes and a `ConstantString` that inherits from this and packs a string into those 8 bytes.

The proposal would add a `VariableLengthConstant` and a `VariableLengthConstantString`. Alternate naming could be `LongConstant` and `LongConstantString`.

FussyDuck commented:
[![CLA assistant check](https://bol-claassistant.ecmwf.int:443/pull/badge/not_signed)](https://bol-claassistant.ecmwf.int:443/ecmwf/pyodc?pullRequest=14) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://bol-claassistant.ecmwf.int:443/ecmwf/pyodc?pullRequest=14) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://bol-claassistant.ecmwf.int:443/check/ecmwf/pyodc?pullRequest=14) it.</sub>

TomHodson commented:
As of [5bbe711](https://github.com/ecmwf/pyodc/pull/14/commits/5bbe71126e59a95a4b1bb40587f3a7c42f7720a7) I've implemented a VariableLengthConstantString codec which seems to work. I added a test that also exercises the other string codecs to make sure that I understood how the internal decode/encode machinery is meant to work. 

I used stream.encodeString for this which internally uses an Int32 for the length field, is this for backwards compatibility because I would have expected an unsigned value to make more sense here.

simondsmart commented on pyodc/codec.py:
```diff
@@ -151,6 +151,66 @@ def from_dataframe(cls, column_name: str, data: pd.Series, data_type: DataType,
     def decode(self, stream):
         return struct.pack("<d", self.min).split(b"\x00", 1)[0].decode("utf-8")
 
+class VariableLengthConstantString(Constant):
```
This is a very long class name.

In the header, this will be converted directly into a string which is used to identify the codec. I note also that "Variable Constant" is a very odd combination in a name.

Can we suggest something cleaner.

simondsmart commented on pyodc/codec.py:
```diff
@@ -151,6 +151,66 @@ def from_dataframe(cls, column_name: str, data: pd.Series, data_type: DataType,
+        universal (name, dtype, codec) part and the
+        addtional codec specific data.
+        """
+        stream.encodeString(self.column_name) # Column name
```
These are already encoded in the base codec type. We should be using the base encode_header function here. i.e. super().encode_header(). Even if that does cause us to encode unnecessary/unusable other values (yuk)

Same in the decode pathway.

simondsmart reviewed (changes requested):
Also, we must not merge these changes until we have corresponding changes in odc.
